pkg/controller/controller_instance.go

```diff
@@ -1441,12 +1480,13 @@ func (c *controller) processServiceInstanceOperationError(instance *v1beta1.Serv
 
 // processProvisionSuccess handles the logging and updating of a
 // ServiceInstance that has successfully been provisioned at the broker.
-func (c *controller) processProvisionSuccess(instance *v1beta1.ServiceInstance, dashboardURL *string) error {
```

Comment:
Why ever update the `ObservedGeneration` in these `processXY` functions? The way I understood it, `ObservedGeneration` should be set when the controller *starts* processing a resource at a specific generation. Thus it would have been set correctly earlier on.

Reply:
That was my understanding as well. I am not keen on us relying on a side effect of clearing the current operation as the mechanism by which we are setting the `ObservedGeneration`. I would like the `ObservedGeneration` to be set explicitly in the reconcileXXXAdd and reconcileXXXUpdate functions.

One thing that is missing with the current approach is setting the `ObservedGeneration` when there is a change to the spec while an operation is in progress. I understand that is not allowed currently by the API validation, but I think we can accommodate that now to cover us in case we do allow for changes in the future.

Reply:
> I am not keen on us relying on a side effect of clearing the current operation as the mechanism by which we are setting the `ObservedGeneration`

I agree. IMO `clearServiceInstanceCurrentOperation` should no longer do any generation updating. We started to hack around it once we introduced aborting orphan mitigation in favor of deprovisioning, and we only made it worse with the `ReconciledGeneration` hack when deleting polling instances. (I can fix the RG problems in a separate PR, though).

I think the `recordStartOfServiceInstanceOperation` function should work for setting `ObservedGeneration`.

Reply:
@kibbles-n-bytes currently the `ObservedGeneration` is updated when the synchronous provisioning has succeeded, but you're right that `recordStartOfServiceInstanceOperation` is probably a better place to do this.

Reply:
> One thing that is missing with the current approach is setting the ObservedGeneration when there is a change to the spec while an operation is in progress.

@staebler do you have suggestions how to address this? the problem is that with async operation in progress `ObservedGeneration` + `Ready:True` is a sign that the operation has succeeded. So if you update the `ObservedGeneration` while the previous generation is still being processed by the broker, once it succeeds the status will be incorrect (i.e. pointing that a newer generation has been successfully reconciled).
Which makes me think that we might still need `ReconciledGeneration` to cover this...

@kibbles-n-bytes WDYT?

Reply:
> One thing that is missing with the current approach is setting the ObservedGeneration when there is a change to the spec while an operation is in progress.

@nilebox I was not as clear as I could have been in describing what I meant by "while an operation is in progress". I meant that there were pending changes that still needed to be reconciled with the broker, not that there was an async operation in progress.

For example, let's say that the user updates the instance to change the parameters, changing the generation of the instance to 2. The controller observes the new generation and sets the observed generation to 2. The controller sets `CurrentOperation` to `Update`. There is some retryable error that prevents the instance from being updated successfully on the broker. Then, the user makes another update to the instance, changing the generation of the instance to 3. Subsequent requests sent to the broker will use the spec from generation 3. However, the controller will not change the observed generation to 2 because `CurrentOperation` was and remains `Update`. Later, a request to the broker is finally successful. If the request was sync, the controller sets the observed generation to 3. If the request was async, the controller does not set the observed generation: it stays 2.

Reply:
Yeah that makes sense. The comment above was written before we discussed the meaning of `ObservedGeneration` vs `LastAttemptedGeneration` in Slack.